### PR TITLE
Robot Simulator (Part 1)

### PR DESCRIPTION
**Pull request by JoelChanZhiYang**
# Description
The robot simulator has far too much code for reviewers to handle so it is split into 2 sections. This section exclude everything ev3. The section includes the initialization of the simulation, setting up of the physics engine and render engine. It can also render static cuboids. 

This looks like a big PR but about half the code is tests. So only about 1000 lines of actual code

## Type of change
Creation of the simulation set up

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
- Unit tests. This section has 100% coverage
image

- Run the code below to see the simulation come up. This requires you to use the [PR 2882](https://github.com/source-academy/frontend/pull/2882) from the frontend to work. 

```
import {
  createCustomPhysics,
  createPhysics,
  createRenderer,
  init_simulation,
  createCuboid,
  createTimer,
  createWorld,
  createWall,
  createFloor,
  addControllerToWorld,
  createRobotConsole,
  saveToContext
} from "robot_simulation";


init_simulation(() => {
    const physics = createPhysics();
    const renderer = createRenderer();
    const timer = createTimer();
    const robot_console = createRobotConsole();
    
    const world = createWorld(physics, renderer, timer, robot_console);
    
    
    const floor = createFloor(physics, renderer);
    addControllerToWorld(floor, world);
    
    const wall = createWall(physics, renderer);
    addControllerToWorld(wall, world);
    
    saveToContext("world", world);
    
    return world;
});

```

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: .husky/.gitignore
Comment (RichDom2185) [inline .husky/.gitignore:1]: Hi, thanks for helping to update the file permissions in order to get the hooks working!

File: src/bundles/robot_simulation/controllers/index.ts
Comment (RichDom2185) [inline src/bundles/robot_simulation/controllers/index.ts:1]: Was this file intentional?

File: src/tabs/RobotSimulation/components/Modal.tsx [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+export const childWrapperStyle: CSSProperties = {
+  display: 'flex',
+  width: '100%',
+  height: '100%',
+  justifyContent: 'center',
+  alignItems: 'center',
+};
+
+export function Modal({ children, isOpen, onClose }: ModalProps) {
```
Comment (RichDom2185) [inline src/tabs/RobotSimulation/components/Modal.tsx:47]: Can we use `React.FC<ModalProps>` typing with a lambda function instead?

File: src/tabs/RobotSimulation/components/Simulation/index.tsx [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  boxShadow: 'inset 0 0 0 1px rgba(255, 255, 255, 0.2)',
+};
+
+export default function SimulationCanvas({
+  context,
+  isOpen,
+}: {
+  context: DebuggerContext;
+  isOpen: boolean;
+}) {
```
Comment (RichDom2185) [inline src/tabs/RobotSimulation/components/Simulation/index.tsx:37]: This is a component as well right? Can we use React.FC typing?

File: src/tabs/RobotSimulation/components/Main.tsx [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import { useState } from 'react';
+import { type DebuggerContext } from '../../../typings/type_helpers';
+import { Modal } from './Modal';
+import SimulationCanvas from './Simulation';
+import TabUi from './TabUi';
+
+export default function Main({ context }: { context:DebuggerContext }): JSX.Element {
```
Comment (RichDom2185) [inline src/tabs/RobotSimulation/components/Main.tsx:7]: Ditto on typing

File: src/bundles/robot_simulation/controllers/index.ts
Comment (RichDom2185) [inline src/bundles/robot_simulation/controllers/index.ts:1]: Is this empty file intentional?

**Review by RichDom2185 — COMMENTED** [review]
Just some preliminary comments, thanks and looking forward to the simulator!

File: .husky/.gitignore
Comment (JoelChanZhiYang) [reply .husky/.gitignore:1]: No worries!

File: src/bundles/robot_simulation/controllers/index.ts
Comment (JoelChanZhiYang) [reply src/bundles/robot_simulation/controllers/index.ts:1]: It will have stuff in Part 2. But you are right. I deleted it for now.

File: src/tabs/RobotSimulation/components/Modal.tsx (hunk 1 above)
Comment (JoelChanZhiYang) [reply src/tabs/RobotSimulation/components/Modal.tsx]: Fixed!

File: src/tabs/RobotSimulation/components/Simulation/index.tsx (hunk 2 above)
Comment (JoelChanZhiYang) [reply src/tabs/RobotSimulation/components/Simulation/index.tsx]: Fixed!

File: src/tabs/RobotSimulation/components/Main.tsx (hunk 3 above)
Comment (JoelChanZhiYang) [reply src/tabs/RobotSimulation/components/Main.tsx]: Fixed!

File: src/bundles/robot_simulation/index.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+/**
+ * Robot simulator for EV3.
+ *
+ * @module robot_simulation
+ * @author Joel Chan
+ */
+
+export {
+  createCustomPhysics,
```
Comment (leeyi45) [inline src/bundles/robot_simulation/index.ts:9]: Are we using pascal case naming for these exported functions? @RichDom2185

Comment (RichDom2185) [reply src/bundles/robot_simulation/index.ts:9]: Oh, I think you meant camelCase not PascalCase, but yes, you're right. Though I don't think we it should block the PR (since it's only Part 1 anyway).

@JoelChanZhiYang in your subsequent parts can you convert the exported functions to snake_case to match the rest of our modules?

File: src/bundles/robot_simulation/engine/Render/helpers/GLTF.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import * as THREE from 'three';
+// eslint-disable-next-line import/extensions
+import { type GLTF, GLTFLoader } from 'three/examples/jsm/loaders/GLTFLoader.js';
+import type { Dimension } from '../../Math/Vector';
+
+type GLTFLoaderOptions = Dimension;
+
+export function loadGLTF(url: string, dimension: GLTFLoaderOptions): Promise<GLTF> {
```
Comment (RichDom2185) [inline src/bundles/robot_simulation/engine/Render/helpers/GLTF.ts:8]: Interesting that `three` uses `GLTF` capitalisation instead of `Gltf` (thus making it look non-standard), but oh well, I guess we can't do anything about it.

File: src/bundles/robot_simulation/controllers/utils/mergeConfig.ts [hunk 6]
```diff
@@ -0,0 +1,12 @@
+import * as _ from 'lodash';
+
+export type RecursivePartial<T> = T extends object ? {
+  [P in keyof T]?: RecursivePartial<T[P]>;
+} : T;
```
Comment (RichDom2185) [inline src/bundles/robot_simulation/controllers/utils/mergeConfig.ts:5]: Can we move this to common utils and perhaps name it `DeepPartial` (to match frontend)?

**Review by RichDom2185 — APPROVED** [review]
LGTM, thanks a lot!

File: src/bundles/robot_simulation/index.ts (hunk 4 above)
Comment (RichDom2185) [reply src/bundles/robot_simulation/index.ts:9]: On second thought, I realise EV3 functions are written as `ev3_getColorSensor` ("camel-case"-like style), so please KIV on this.

File: src/bundles/robot_simulation/controllers/utils/mergeConfig.ts (hunk 6 above)
Comment (JoelChanZhiYang) [reply src/bundles/robot_simulation/controllers/utils/mergeConfig.ts]: Fixed!
